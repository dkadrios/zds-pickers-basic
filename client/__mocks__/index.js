export inputControls from './inputControls'
export mappings from './mappings'
export midi from './midi'
export shiftGroups from './shiftGroups'
export shifter from './shifter'
export store from './store'
export user from './user'
export version from './version'
