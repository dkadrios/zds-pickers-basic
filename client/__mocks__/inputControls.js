export default [
  {
    controlId: 0,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 11,
    value: 110,
    active: true,
    lit: true,
  },
  {
    controlId: 1,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 1,
    status: 11,
    value: 111,
    active: false,
    lit: false,
  },
  {
    controlId: 2,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: false,
  },
  {
    controlId: 3,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: true,
  },
  {
    controlId: 4,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: true,
  },
  {
    controlId: 5,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: false,
  },
]
