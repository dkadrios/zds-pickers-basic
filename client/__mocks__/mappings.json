{
  "dialogVisible": false,
  "userDialogVisible": false,
  "userImportDialogVisible": false,
  "channels": [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    "BFD3",
    null,
    null,
    null,
    null,
    null,
    null
  ],
  "stockMappings": [
    "2box",
    "Addictive Drums",
    "Alesis DM Pro",
    "Alesis DM10",
    "Alesis DM5",
    "BFD2",
    "BFD3",
    "Battery 3",
    "E-MU Protean Acoustic",
    "E-MU Protean Bread n Butter",
    "E-MU Protean Studio",
    "E-MU Protean Synth",
    "EZ Drummer The Classic EZX",
    "EZDrummer",
    "EZDrummer Jazz Kit",
    "EZDrummer Latin Percussion",
    "General MIDI",
    "Infinite Instruments",
    "John Emrich - Articulations 1",
    "Kontakt 2",
    "Ocean Way Drums",
    "Pearl Redbox",
    "Roland HandSonic 15",
    "Roland TD-10",
    "Roland TD-10 Perc 1",
    "Roland TD-10 Perc 2",
    "Roland TD-10 Perc 3",
    "Roland TD-10 Perc 4",
    "Roland TD-12",
    "Roland TD-12 Perc 1 (Latin)",
    "Roland TD-12 Perc 2 (Indian)",
    "Roland TD-12 Perc 3 (African)",
    "Roland TD-12 Perc 4 (Salsa)",
    "Roland TD-12 Perc 5 (SFX)",
    "Roland TD-12 Perc 6-8 (User)",
    "Roland TD-20",
    "Roland TD-20 Perc",
    "Roland TD-3",
    "Roland TD-30",
    "Roland TD-30 Perc 1 (Latin)",
    "Roland TD-30 Perc 2 (Indian)",
    "Roland TD-30 Perc 3 (African)",
    "Roland TD-30 Perc 4 (Salsa)",
    "Roland TD-30 Perc 5 (SFX)",
    "Roland TD-30 Perc 6-8 (User)",
    "Roland TD-5",
    "Roland TD-6",
    "Roland TD-6 Perc 1",
    "Roland TD-6 Perc 10 (Special)",
    "Roland TD-6 Perc 2",
    "Roland TD-6 Perc 3 (Room)",
    "Roland TD-6 Perc 4 (Power)",
    "Roland TD-6 Perc 5 (Elec)",
    "Roland TD-6 Perc 6 (808)",
    "Roland TD-6 Perc 7 (Jazz)",
    "Roland TD-6 Perc 8 (Brush)",
    "Roland TD-6 Perc 9 (Perc)",
    "Roland TD-7",
    "Roland TD-8",
    "Roland TD-8 Perc 1",
    "Roland TD-8 Perc 10 (Special)",
    "Roland TD-8 Perc 2",
    "Roland TD-8 Perc 3 (Room)",
    "Roland TD-8 Perc 4 (Power)",
    "Roland TD-8 Perc 5 (Elec)",
    "Roland TD-8 Perc 6 (808)",
    "Roland TD-8 Perc 7 (Jazz)",
    "Roland TD-8 Perc 8 (Brush)",
    "Roland TD-8 Perc 9 (Perc)",
    "Roland TD-9",
    "Roland TD-9 Perc",
    "Simmons SD1000",
    "Simmons SD5X",
    "Simmons SD7",
    "Simmons SD7PK",
    "Simmons SD9K",
    "Sonic Reality iMap Extended",
    "Sonic Reality iMap Standard",
    "Studio ProFiles",
    "Superior Drummer",
    "Yamaha DTX 700",
    "Yamaha DTXplorer",
    "Zendrum Stompblock",
    "Zendrum TRS",
    "Zendrum TRS 2.0"
  ],
  "userMappings": ["BFD3 Custom ZenEdit", "reaper"]
}
