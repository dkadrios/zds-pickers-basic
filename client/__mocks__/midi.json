{
  "devices": [
    {
      "id": "-1325431060",
      "manufacturer": "Zendrum Studio",
      "name": "ZDS Shifter",
      "type": "output",
      "version": "",
      "state": "connected",
      "connection": "open"
    },
    {
      "id": "-312145282",
      "manufacturer": "Zendrum Studio",
      "name": "ZDS Shifter",
      "type": "input",
      "version": "",
      "state": "connected",
      "connection": "open"
    }
  ],
  "listeningDevices": ["-312145282"]
}
