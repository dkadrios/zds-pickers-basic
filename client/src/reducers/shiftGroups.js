import { assertRange } from 'zds-pickers'
import { arraySequence, chunk, splitAt } from '../core/fp/arrays'
import { sum } from '../core/fp/numbers'
import { GROUP_SIZE_BYTES, MAX_ENTRIES, MAX_GROUPS } from '../core/consts'
import { createReducer } from './utils'
import shiftGroup from './shiftGroup'
import actionTypes from './actionTypes.js'

export const actions = {
  askForGroups: () => ({
    type: actionTypes.GET_SYSEX_GROUPS,
  }),

  receivedGroups: groupData => ({
    type: actionTypes.RECEIVED_GROUPS,
    groupData,
  }),

  changeSelectedGroup: groupId => ({
    type: actionTypes.CHANGE_SELECTED_GROUP,
    groupId,
  }),

  changeGroupChannel: (groupId, channel) => ({
    type: actionTypes.CHANGE_GROUP_CHANNEL,
    groupId,
    channel: assertRange(channel, 15, 0),
  }),

  changeGroupValue: (groupId, value) => ({
    type: actionTypes.CHANGE_GROUP_VALUE,
    groupId,
    value: assertRange(value, 127, 0),
  }),

  editEntry: (groupId, editEntryIdx) => ({
    type: actionTypes.EDIT_ENTRY,
    groupId,
    editEntryIdx,
  }),

  addEntry: groupId => ({
    type: actionTypes.ADD_ENTRY,
    groupId,
  }),

  saveEntryEdit: (groupId, editQueue) => ({
    type: actionTypes.SAVE_ENTRY_EDIT,
    groupId,
    editQueue,
  }),

  cancelEntryEdit: groupId => ({
    type: actionTypes.CANCEL_ENTRY_EDIT,
    groupId,
  }),

  removeEntry: (groupId, entryId) => ({
    type: actionTypes.REMOVE_ENTRY,
    groupId,
    entryId,
  }),

  changeSort: (groupId, sortOn, sortBy) => ({
    type: actionTypes.SORT_GROUP,
    groupId,
    sortOn,
    sortBy,
  }),

  selectShiftEntry: (groupId, selectedEntryId) => ({
    type: actionTypes.SELECT_SHIFT_ENTRY,
    groupId,
    selectedEntryId,
  }),

  selectAllEntries: groupId => ({
    type: actionTypes.SELECT_ALL_ENTRIES,
    groupId,
  }),

  clearAllSelections: groupId => ({
    type: actionTypes.CLEAR_ALL_SELECTIONS,
    groupId,
  }),

  handleEntryClick: (groupId, sortedIds, idx, meta, shift) => ({
    type: actionTypes.ENTRY_META_CLICKED,
    groupId,
    sortedIds,
    idx,
    meta,
    shift,
  }),
}

const defaultState = {
  selectedGroupIdx: 0,
  maxEntries: MAX_ENTRIES,
  totalEntries: 0,
  groups: [],
}

const passThrough = (state, action) => ({
  ...state,

  groups: state.groups.map(group => (group.groupId === action.groupId ? shiftGroup(group, action) : { ...group })),
})

const changeSelectedGroup = (state, { groupId }) => ({
  ...state,
  selectedGroupIdx: groupId,
})

const addRemoveEntry = (state, action) => {
  const newState = { ...passThrough(state, action) }
  return {
    ...newState,
    totalEntries: newState.groups.reduce((acc, group) => acc + group.entries.length, 0),
  }
}

const handleShifterUnplugged = () => ({
  ...defaultState,
})

const receivedGroups = (state, { groupData }) => {
  const [preamble, rest] = splitAt(MAX_GROUPS * GROUP_SIZE_BYTES)(groupData)

  const [numEntries, channels, values] = chunk(MAX_GROUPS)(preamble)

  const groups = arraySequence(MAX_GROUPS).map((val, groupId) => ({
    groupId,
    channel: channels[groupId],
    value: values[groupId],
    label: `Group ${String.fromCharCode(65 + groupId)}`,
  }))

  return {
    ...state,
    totalEntries: sum(...numEntries),
    groups: groups.map((
      group,
      idx, //
    ) => shiftGroup(group, {
      type: actionTypes.RECEIVED_GROUP, //
      entries: rest.splice(0, numEntries[idx] * MAX_GROUPS),
    })),
  }
}

const receivedInternalState = (state, { packet }) => {
  const [numControls, ...rest] = packet
  rest.pop() // Settings flag

  const active = rest.slice(numControls * 2)

  return {
    ...state,
    groups: state.groups.map((group, idx) => ({ ...group, active: Boolean(active[idx]) })),
  }
}

export default createReducer(defaultState, {
  [actionTypes.CHANGE_SELECTED_GROUP]: changeSelectedGroup,
  [actionTypes.CHANGE_GROUP_CHANNEL]: passThrough,
  [actionTypes.CHANGE_GROUP_VALUE]: passThrough,
  [actionTypes.ADD_ENTRY]: passThrough,
  [actionTypes.EDIT_ENTRY]: passThrough,
  [actionTypes.SAVE_ENTRY_EDIT]: addRemoveEntry,
  [actionTypes.CANCEL_ENTRY_EDIT]: passThrough,
  [actionTypes.REMOVE_ENTRY]: addRemoveEntry,
  [actionTypes.CHANGE_ENTRY_STATUS]: passThrough,
  [actionTypes.CHANGE_ENTRY_CHANNEL]: passThrough,
  [actionTypes.CHANGE_ENTRY_VALUE]: passThrough,
  [actionTypes.SHIFTER_MISSING]: handleShifterUnplugged,
  [actionTypes.FACTORY_RESET]: handleShifterUnplugged,
  [actionTypes.RECEIVED_GROUPS]: receivedGroups,
  [actionTypes.SORT_GROUP]: passThrough,
  [actionTypes.SELECT_SHIFT_ENTRY]: passThrough,
  [actionTypes.RECEIVED_STATE]: receivedInternalState,
  [actionTypes.SELECT_ALL_ENTRIES]: passThrough,
  [actionTypes.CLEAR_ALL_SELECTIONS]: passThrough,
  [actionTypes.ENTRY_META_CLICKED]: passThrough,
})
