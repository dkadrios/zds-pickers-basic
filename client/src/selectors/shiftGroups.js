import { createSelector } from 'reselect'
import { takeSecond } from '../core/fp/utils'
import { findObj } from '../core/fp/arrays'
import { isDisabled } from './shifter'
import { stateInputControls, stateMappings, stateShiftGroups, stateShifter, stateUser } from '.'

export const getShiftGroups = createSelector(
  stateShiftGroups,
  stateInputControls,
  (shiftGroups, inputControls) => {
    const groups = shiftGroups?.groups || []
    const decoratedGroups = groups.map((group) => {
      const assignedControls = inputControls
        .filter(({ channel, value }) => channel === group.channel && value === group.value)
        .map(({ controlId }) => controlId + 1)
        .map(controlNumber => `FS${controlNumber}`)
      return { ...group, controlLabels: assignedControls }
    })

    return {
      ...shiftGroups,
      groups: decoratedGroups,
    }
  },
)

export const getShiftGroup = createSelector(
  getShiftGroups,
  takeSecond,
  ({ groups }, groupId) => findObj('groupId', groupId)(groups),
)

export const canAddEntryToShiftGroup = createSelector(
  getShiftGroup,
  stateMappings,
  stateShifter,
  stateShiftGroups,
  stateUser,
  isDisabled,
  (shiftGroup, mappings, shifter, shiftGroups, user, disabled) => {
    const {
      errorVisible,
      exportDialogVisible,
      found,
      importDialogVisible,
      importInProcess,
      responding,
    } = shifter

    const { dialogVisible } = user

    const { editing } = shiftGroup || {}

    const {
      maxEntries,
      totalEntries,
    } = shiftGroups

    const {
      userDialogVisible,
      userImportDialogVisible,
    } = mappings

    return responding
    && found
    && !disabled
    && !editing
    && !dialogVisible
    && !importDialogVisible
    && !importInProcess
    && !userDialogVisible
    && !userImportDialogVisible
    && !exportDialogVisible
    && !errorVisible
    // && !registrationDialogVisible // TODO:
    && totalEntries < maxEntries
  },
)

export default undefined
