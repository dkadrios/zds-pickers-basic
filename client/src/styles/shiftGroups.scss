@import 'variables';
@import 'mixins';

:local {
  .container {
    border: 4px solid $color-shifter-red;
    border-radius: 0 4px 4px 4px;
  }

  .loadingCont {
    padding-top: 80px;
    text-align: center;
    font-size: 24px;
    color: $color-text-dimmer;
    height: calc(100vh - #{$grid-height} + 55px) !important;
    overflow: hidden !important;
  }
}
