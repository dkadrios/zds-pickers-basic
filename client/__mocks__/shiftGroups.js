import { SORT_ASC, SORT_BY_ALL, SORT_ON_INPUT } from '../src/reducers/shiftGroup'

export const groupAEntryData = [153, 1, 153, 1, 153, 2, 153, 2, 153, 3, 153, 3, 153, 4, 153, 4]

export const entryData = [
  ...groupAEntryData,
  153,
  5,
  153,
  5,
  153,
  6,
  153,
  6,
  153,
  7,
  153,
  7,
  153,
  8,
  153,
  8,
  153,
  9,
  153,
  9,
  153,
  10,
  153,
  10,
]

export const groupData = [4, 3, 2, 1, 9, 9, 9, 9, 110, 111, 112, 113, ...entryData]

export const group = {
  channel: 9,
  editQueue: {},
  editing: false,
  selectedEntryId: NaN,
  selectedRows: [],
  active: false,
  entries: [
    {
      entryId: 0,
      input: { channel: 9, status: 9, value: 1 },
      output: { channel: 9, status: 9, value: 1 },
    },
    {
      entryId: 1,
      input: { channel: 9, status: 9, value: 2 },
      output: { channel: 9, status: 9, value: 2 },
    },
    {
      entryId: 2,
      input: { channel: 9, status: 9, value: 3 },
      output: { channel: 9, status: 9, value: 3 },
    },
    {
      entryId: 3,
      input: { channel: 9, status: 9, value: 4 },
      output: { channel: 9, status: 9, value: 4 },
    },
  ],
  groupId: 0,
  label: 'Group A',
  value: 110,
  sortOn: SORT_ON_INPUT,
  sortBy: SORT_BY_ALL,
  sortDir: SORT_ASC,
}

export const groups = {
  groups: [
    {
      channel: 9,
      editQueue: {
        entryId: -1,
        input: { channel: 9, status: 9, value: 0 },
        output: { channel: 9, status: 9, value: 0 },
      },
      editing: false,
      entries: [
        {
          entryId: 0,
          input: { channel: 9, status: 9, value: 1 },
          output: { channel: 9, status: 9, value: 1 },
        },
        {
          entryId: 1,
          input: { channel: 9, status: 9, value: 2 },
          output: { channel: 9, status: 9, value: 2 },
        },
        {
          entryId: 2,
          input: { channel: 9, status: 9, value: 3 },
          output: { channel: 9, status: 9, value: 3 },
        },
        {
          entryId: 3,
          input: { channel: 9, status: 9, value: 4 },
          output: { channel: 9, status: 9, value: 4 },
        },
      ],
      sortOn: SORT_ON_INPUT,
      sortBy: SORT_BY_ALL,
      sortDir: SORT_ASC,
      groupId: 0,
      label: 'Group A',
      value: 110,
      selectedEntryId: NaN,
      selectedRows: [],
      active: false,
    },
    {
      channel: 9,
      editQueue: {
        entryId: -1,
        input: { channel: 9, status: 9, value: 0 },
        output: { channel: 9, status: 9, value: 0 },
      },
      editing: false,
      entries: [
        {
          entryId: 0,
          input: { channel: 9, status: 9, value: 5 },
          output: { channel: 9, status: 9, value: 5 },
        },
        {
          entryId: 1,
          input: { channel: 9, status: 9, value: 6 },
          output: { channel: 9, status: 9, value: 6 },
        },
        {
          entryId: 2,
          input: { channel: 9, status: 9, value: 7 },
          output: { channel: 9, status: 9, value: 7 },
        },
      ],
      sortOn: SORT_ON_INPUT,
      sortBy: SORT_BY_ALL,
      sortDir: SORT_ASC,
      groupId: 1,
      label: 'Group B',
      value: 111,
      selectedEntryId: NaN,
      selectedRows: [],
      active: false,
    },
    {
      channel: 9,
      editQueue: {
        entryId: -1,
        input: { channel: 9, status: 9, value: 0 },
        output: { channel: 9, status: 9, value: 0 },
      },
      editing: false,
      entries: [
        {
          entryId: 0,
          input: { channel: 9, status: 9, value: 8 },
          output: { channel: 9, status: 9, value: 8 },
        },
        {
          entryId: 1,
          input: { channel: 9, status: 9, value: 9 },
          output: { channel: 9, status: 9, value: 9 },
        },
      ],
      sortOn: SORT_ON_INPUT,
      sortBy: SORT_BY_ALL,
      sortDir: SORT_ASC,
      groupId: 2,
      label: 'Group C',
      value: 112,
      selectedEntryId: NaN,
      selectedRows: [],
      active: false,
    },
    {
      channel: 9,
      editQueue: {
        entryId: -1,
        input: { channel: 9, status: 9, value: 0 },
        output: { channel: 9, status: 9, value: 0 },
      },
      editing: false,
      entries: [
        {
          entryId: 0,
          input: { channel: 9, status: 9, value: 10 },
          output: { channel: 9, status: 9, value: 10 },
        },
      ],
      sortOn: SORT_ON_INPUT,
      sortBy: SORT_BY_ALL,
      sortDir: SORT_ASC,
      groupId: 3,
      label: 'Group D',
      value: 113,
      selectedEntryId: NaN,
      selectedRows: [],
      active: false,
    },
  ],
  maxEntries: 112,
  selectedGroupIdx: 0,
  totalEntries: 10,
}

export default groups
