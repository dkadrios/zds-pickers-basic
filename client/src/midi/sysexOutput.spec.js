import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import backup from '../../__assets__/zds-shifter-backup.json'
import { group } from '../../__mocks__/shiftGroups'
import {
  askForBackup,
  askForControls,
  askForGroups,
  askForModel,
  askForVersion,
  performFactoryReset,
  removeShiftEntry,
  restart,
  saveShiftEntry,
  setFlags,
  transmitControls,
  transmitShiftGroupChannel,
  transmitShiftGroupValue,
} from './sysexOutput'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)
const store = mockStore({})

describe('sysexOutput processing', () => {
  beforeEach(store.clearActions)

  it('askForVersion should work', () => {
    askForVersion(store.dispatch, '1234567890')
    expect(store.getActions()).toEqual([
      {
        payload: {
          data: [240, 108, 0, 1, 49, 50, 51, 52, 53, 54, 0, 55, 56, 57, 48, 247],
          device: -1,
          timestamp: expect.any(Number),
        },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })

  it('askForModel should work', () => {
    askForModel(store.dispatch)
    expect(store.getActions()).toEqual([
      {
        payload: {
          data: [240, 108, 0, 7, 247],
          device: -1,
          timestamp: expect.any(Number),
        },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })

  it('askForControls should work', () => {
    askForControls(store.dispatch)
    expect(store.getActions()).toEqual([
      {
        payload: {
          data: [240, 108, 0, 3, 247],
          device: -1,
          timestamp: expect.any(Number),
        },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })

  it('transmitControls should work', () => {
    transmitControls(store.dispatch, backup.inputControls)
    expect(store.getActions()).toMatchSnapshot()
  })

  it('askForGroups should work', () => {
    askForGroups(store.dispatch)
    expect(store.getActions()).toEqual([
      {
        payload: {
          data: [240, 108, 0, 5, 247],
          device: -1,
          timestamp: expect.any(Number),
        },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })

  it('restart should work', () => {
    restart(store.dispatch)
    expect(store.getActions()).toEqual([
      {
        payload: {
          data: [240, 108, 0, 125, 247],
          device: -1,
          timestamp: expect.any(Number),
        },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })

  it('performFactoryReset should work without restart', () => {
    performFactoryReset(store.dispatch, false)
    expect(store.getActions()).toEqual([
      {
        payload: {
          data: [240, 108, 0, 126, 0, 247],
          device: -1,
          timestamp: expect.any(Number),
        },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })

  it('performFactoryReset should work with restart', () => {
    performFactoryReset(store.dispatch, true)
    expect(store.getActions()).toEqual([
      {
        payload: {
          data: [240, 108, 0, 126, 1, 247],
          device: -1,
          timestamp: expect.any(Number),
        },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })

  it('setFlags', () => {
    setFlags(store.dispatch, { midiActivityLEDMode: 3, serialMidiOutEnabled: true, usbMidiOutEnabled: false })
    expect(store.getActions()).toEqual([
      {
        payload: { data: [240, 108, 0, 10, 7, 247], device: -1, timestamp: 0 },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])

    store.clearActions()
    setFlags(store.dispatch, { midiActivityLEDMode: 3, serialMidiOutEnabled: false, usbMidiOutEnabled: true })
    expect(store.getActions()).toEqual([
      {
        payload: { data: [240, 108, 0, 10, 11, 247], device: -1, timestamp: 0 },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })

  it('transmitShiftGroupChannel', () => {
    transmitShiftGroupChannel(store.dispatch, { groupId: 1, channel: 9 })
    expect(store.getActions()).toEqual([
      {
        payload: { data: [240, 108, 0, 11, 1, 9, 247], device: -1, timestamp: 0 },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })

  it('transmitShiftGroupValue', () => {
    transmitShiftGroupValue(store.dispatch, { groupId: 1, channel: 9 })
    expect(store.getActions()).toEqual([
      {
        payload: { data: [240, 108, 0, 12, 1, 0, 247], device: -1, timestamp: 0 },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })

  it('removeShiftEntry - single entry', () => {
    removeShiftEntry(store.dispatch, { groupId: 1, entryId: 9 })
    expect(store.getActions()).toEqual([
      {
        payload: { data: [240, 108, 0, 14, 1, 9, 247], device: -1, timestamp: 0 },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })

  it('removeShiftEntry - multiple entries', () => {
    removeShiftEntry(store.dispatch, { groupId: 1, entryId: [3, 9] })
    expect(store.getActions()).toMatchSnapshot()
  })

  it('saveShiftEntry', () => {
    saveShiftEntry(store.dispatch, { groupId: 0, editQueue: group.entries[0] })
    expect(store.getActions()).toMatchSnapshot()
    store.clearActions()

    saveShiftEntry(store.dispatch, { groupId: 0, editQueue: { ...group.entries[1], entryId: -1 } })
    expect(store.getActions()).toMatchSnapshot()
  })

  it('askForBackup', () => {
    askForBackup(store.dispatch)
    expect(store.getActions()).toEqual([
      {
        payload: { data: [240, 108, 0, 16, 247], device: -1, timestamp: 0 },
        type: 'redux-midi/midi/SEND_MIDI_MESSAGE',
      },
    ])
  })
})
