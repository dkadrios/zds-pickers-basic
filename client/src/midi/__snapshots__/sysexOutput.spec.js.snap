// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sysexOutput processing removeShiftEntry - multiple entries 1`] = `
Array [
  Object {
    "payload": Object {
      "data": Array [
        240,
        108,
        0,
        14,
        1,
        3,
        247,
      ],
      "device": -1,
      "timestamp": 0,
    },
    "type": "redux-midi/midi/SEND_MIDI_MESSAGE",
  },
  Object {
    "payload": Object {
      "data": Array [
        240,
        108,
        0,
        14,
        1,
        9,
        247,
      ],
      "device": -1,
      "timestamp": 0,
    },
    "type": "redux-midi/midi/SEND_MIDI_MESSAGE",
  },
]
`;

exports[`sysexOutput processing saveShiftEntry 1`] = `
Array [
  Object {
    "payload": Object {
      "data": Array [
        240,
        108,
        40,
        13,
        0,
        0,
        25,
        1,
        25,
        1,
        247,
      ],
      "device": -1,
      "timestamp": 0,
    },
    "type": "redux-midi/midi/SEND_MIDI_MESSAGE",
  },
]
`;

exports[`sysexOutput processing saveShiftEntry 2`] = `
Array [
  Object {
    "payload": Object {
      "data": Array [
        240,
        108,
        44,
        13,
        0,
        127,
        25,
        2,
        25,
        2,
        247,
      ],
      "device": -1,
      "timestamp": 0,
    },
    "type": "redux-midi/midi/SEND_MIDI_MESSAGE",
  },
]
`;

exports[`sysexOutput processing transmitControls should work 1`] = `
Array [
  Object {
    "payload": Object {
      "data": Array [
        240,
        108,
        66,
        4,
        58,
        110,
        0,
        0,
        127,
        58,
        16,
        111,
        1,
        0,
        127,
        26,
        110,
        0,
        4,
        0,
        127,
        26,
        110,
        0,
        0,
        127,
        33,
        26,
        110,
        0,
        0,
        127,
        26,
        110,
        0,
        0,
        0,
        127,
        247,
      ],
      "device": -1,
      "timestamp": 0,
    },
    "type": "redux-midi/midi/SEND_MIDI_MESSAGE",
  },
]
`;
